src/psyclone/psyir/symbols/symbol.py

```diff
@@ -244,6 +244,115 @@ def __init__(self, name, visibility=DEFAULT_VISIBILITY, interface=None):
+        :type args: list of obj
+        :param kwargs: optional arguments to provide to the constructor.
+        :type kwargs: dict of keyword-value pairs.
+        :param interface: optional value with which to override this object's \
+                interface in the new object.
+        :type interface: :py:class:`psyclone.psyir.symbols.SymbolInterface`
+        :param visibility: optional value with which to override this object's\
+                visibility in the new object.
+        :type visibility: :py:class:`psyclone.psyir.symbols.Symbol.Visibility`
+
+        :returns: A symbol object with the same properties as this \
+                  symbol object.
+        :rtype: :py:class:`psyclone.psyir.symbols.Symbol` or subclass
+
+        '''
+        if interface:
+            kwargs["interface"] = interface
+        else:
+            kwargs["interface"] = self.interface
+        if visibility:
+            kwargs["visibility"] = visibility
+        else:
+            kwargs["visibility"] = self.visibility
+        # The constructors for all Symbol-based classes have 'name' as the
+        # first positional argument.
+        return type(self)(self.name, *args, **kwargs)
+
+    def get_external_symbol(self):
+        '''
+        Looks-up and returns the Symbol referred to by this Symbol's external
+        (global) interface.
+
+        :raises SymbolError: if the module pointed to by the symbol interface \
+                             does not contain the symbol (or the symbol is \
+                             not public).
+        :raises NotImplementedError: if the this symbol does not have an \
+                                     external (global) interface.
+        '''
+        if not self.is_global:
+            raise NotImplementedError(
+                "Error trying to resolve symbol '{0}' properties, the lazy"
+                " evaluation of '{1}' interfaces is not supported."
+                "".format(self.name, self.interface))
+
+        module = self.interface.container_symbol
+        try:
+            return module.container.symbol_table.lookup(
+                self.name, visibility=self.Visibility.PUBLIC)
+        except KeyError as kerr:
+            six.raise_from(SymbolError(
+                "Error trying to resolve the properties of symbol "
+                "'{0}'. The interface points to module '{1}' but "
+                "could not find the definition of '{0}' in that "
+                "module.".format(self.name, module.name)), kerr)
+        except SymbolError as err:
+            six.raise_from(SymbolError(
+                "Error trying to resolve the properties of symbol "
+                "'{0}' in module '{1}': {2}".format(
+                    self.name, module.name, str(err.value))), err)
+
+    def resolve_deferred(self):
+        '''
+        Search for the Container in which this Symbol is defined and
+        create and return a symbol of the correct class and type. If this
```

Comment:
As I said, I like this implementation now but since we slightly modify the functionality in a specialization of this method I would mention this in the generic docstring. For instance, by adding after the first sentence: "If the class and type are the same, some specializations of this method update the differing properties in-place."

Reply:
Done.